In [33]:
from IPython.core.display import HTML
HTML('<a href="https://colab.research.google.com/github/yongfu-lu/CSC_448_HW3/blob/main/HW3.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a >')

In [11]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures
import matplotlib.pyplot as plt
import numpy as np

# Part 1. Loading Dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
data = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")

In [3]:
# Output the first 15 rows of the data
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [4]:
# Display a summary of the table information (number of datapoints, etc.)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


# Part 2. Splitting Dataset

In [5]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data.drop(columns=['Size nm^3'])
y = data['Size nm^3']

In [6]:
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1)

In [7]:
X_train

Temperature °C  Mols KCL
337             127        26
72              234       698
872             254       662
107             675       759
382             727       839
..              ...       ...
783             732       156
848             772       382
478             448       159
770             114       939
417             696       193

[900 rows x 2 columns]

In [8]:
X_test

Temperature °C  Mols KCL
985              16        53
447             592       739
993             311       265
463             260       530
878             193       316
..              ...       ...
217              91       467
509              31       127
55              965       721
868             216       713
747             642       875

[100 rows x 2 columns]

# Part 3. Perform a linear regression

In [9]:
# Use sklearn to train a model on the training set
model = LinearRegression().fit(X_train, y_train)

In [12]:
# Create a sample datapoint and predict the output of that sample with the trained model
sample_x = np.array([[500, 500]])
sample_y = model.predict([[500, 500]])[0]
print(f"Predict output is: {sample_y:.5f}")

Predict output is: 537700.76548


In [15]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
score_of_traning = model.score(X_train, y_train)
score_of_testing = model.score(X_test, y_test)
print(f"Score of training : {score_of_traning:.5f} \nScore of testing : {score_of_testing:.5f}.")

Score of training : 0.86113 
Score of testing : 0.85547.


The score measures the corelation between the input and output value. The absolute value of the score can vary from 0 to 1, and it can be positive or negative depends on the relationship between the input and output. Score value of 0 means no relation at all. Score value of 1 means a perfect correlation. 

In our above case, the input and output variable of both training dataset and testing dataset has the positive correlation.

In [19]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
co1 = round(model.coef_[0], 5)
co2 = round(model.coef_[1], 5)
intercept_point = round(model.intercept_, 5)
print(f"The coefficents are {co1} and {co2} \nThe intercept point is {intercept_point}")

The coefficents are 881.46942 and 1033.73574 
The intercept point is -419901.81312


Equation for h(x):

$h(x) = 881.46942x_1 + 1033.73574x_2 - 419901.81312$,  
$x_1$ is the Temperature and $x_2$ is the Mols KCL

# Part 4. Use cross validation

In [21]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cross_validations = cross_val_score(model, X_train, y_train, cv = 10)
print(f' Score for each time: {cross_validations}')

 Score for each time: [0.87537519 0.86925467 0.84121183 0.88493961 0.84186229 0.8810148
 0.86625849 0.83887556 0.8587267  0.83516287]


In [22]:
print(f"The average score: {k_folds.mean()}")

The average score: 0.8592681985591175


In [ ]:
# Report on their finding and their significance

The method cross_val_score will return a list of score. The size of the list is indicated by the argument cv. In the above case, the size of the list is 10. Which means the input training data was splitted into 10 parts, and the returned list contains the score coresponding to each part. 

The cross validation provide the way to measure the accuracy of the model by spliting training data into many parts and comparing the result amoung them. It does not required validation set, therefore it is very efficient for a small size data.

# Part 5. Use Polynomial regression 

In [27]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
poly_x_train = poly.fit_transform(X_train)
poly_x_test = poly.fit_transform(X_test)
model_2 = LinearRegression().fit(poly_x_train, y_train)

In [28]:
# Report on the metrics and output the resultant equation as you did in Part 3.
train_score = model_2.score(poly_x_train, y_train)
test_score = model_2.score(poly_x_test, y_test)
print(f"Train score is {train_score:.5f} and Test Score is {test_score:.5f}.")

Train score is 1.00000 and Test Score is 1.00000.


By using polynomial regression, the score of training and testing both increase to 1.0. This indicates a perfect correlation between input and output value.

In [32]:
cos = np.round_(model_2.coef_, decimals = 5)
intercept = model_2.intercept_
print(f"The coefficients are : {cos} \nThe intercept is: {intercept:.5f}")

The coefficients are : [ 0.      12.      -0.      -0.       2.       0.02857] 
The intercept is: 0.00002


Equation is: 
$y = 0.00002 + 12x_1 + 2x_1x_2 + 0.02857x_2^2$,  
$x_1$ is the Temperature and $x_2$ is the Mols KCL.